scli-runner/src/main/scala/com.olegych.scastie.sclirunner/ScliRunner.scala

```diff
@@ -58,45 +70,85 @@ class ScliRunner {
-        val allDirectives = (runtimeDependency.map(scalaDepToFullName).map(libraryDirective) ++ userDirectives)
-        val totalOffset = -runtimeDependency.size + Instrument.getExceptionLineOffset(task.inputs)
+      case Right(InstrumentedInputs(inputs, isForcedProgramMode)) =>
+        buildAndRun(task.snippetId, inputs, isForcedProgramMode, userDirectives, userCode, onOutput)
+    }
+  }
+
+  def buildAndRun(snippetId: SnippetId, inputs: Inputs, isForcedProgramMode: Boolean, userDirectives: Array[String], userCode: Array[String], onOutput: String => Any): Future[ScliRun] = {
+    val runtimeDependency = inputs.target.runtimeDependency.map(Set(_)).getOrElse(Set()) ++ inputs.libraries
+    val allDirectives = (runtimeDependency.map(scalaDepToFullName).map(libraryDirective) ++ userDirectives)
+    val totalOffset = -runtimeDependency.size + Instrument.getExceptionLineOffset(inputs)
+
+    val charOffsetInstrumentation = userDirectives.map(_.length() + 1).sum
 
-        val charOffsetInstrumentation = userDirectives.map(_.length() + 1).sum
+    val code = allDirectives.mkString("\n") + "\n" + inputs.code
+    writeFile(scalaMain, code)
 
-        val code = allDirectives.mkString("\n") + "\n" + inputs.code
-        writeFile(scalaMain, code)
+    var instrumentationMem: Option[List[Instrumentation]] = None
+    var outputBuffer: List[String] = List()
 
-        var instrumentationMem: Option[List[Instrumentation]] = None
+    val build = bspClient.build(snippetId.base64UUID)
+    build.map { bspBuild =>
+      log.info(s"Running process for snippet $snippetId")
 
-        val build = bspClient.build(task.snippetId.base64UUID, line => {
-          extract[List[Instrumentation]](line) match { // extract instrumentation
-            case None => onOutput(line)
+      def forwardPrint(str: String) = {
+        onOutput(str)
+        outputBuffer = str :: outputBuffer
+      }
+        
+      val runProcess = bspBuild.process.run(
+        ProcessLogger({ line: String => {
+          // extract instrumentation
+          extract[List[Instrumentation]](line) match {
+            case None => forwardPrint(line)
             case Some(value) => {
               instrumentationMem = Some(value.map(inst => inst.copy(
                 position = inst.position.copy(inst.position.start + charOffsetInstrumentation, inst.position.end + charOffsetInstrumentation)
               )))
             }
           }
-        })
-
-        build map { r =>
-          r.copy(instrumentation = instrumentationMem)
-        } recover {
-          case x: BspClient.CompilationError => throw CompilationError(x.toProblemList.map(pb =>
-            pb.copy(line = pb.line.map(_ + totalOffset + 1))))
-          case other => throw other
-        } 
+
+        }})
+      )
+      javaProcesses.put(snippetId, runProcess)
+
+      // Wait for process to exit gracefully (or being killed by timeout…)
+      val f = Future { runProcess.exitValue() }
```

Comment:
Yeah, Scala doesn't provide a good way to timeout runs.

You may try to use https://doc.akka.io/docs/akka/current/futures.html#after as we have Akka as dependency anyway, and when we will be migrating away from it we will probably do to some other library providing such possibility eg. cats-effect / monix.

Check if akka's `after` can be used to implement this. If so change it, if not leave it for now as it is.